sdks/java/io/amqp/src/main/java/org/apache/beam/sdk/io/amqp/AmqpIO.java

```diff
@@ -0,0 +1,458 @@
+      this.current = null;
+      if (checkpointMark != null) {
+        this.checkpointMark = checkpointMark;
+      } else {
+        this.checkpointMark = new AmqpCheckpointMark();
+      }
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return checkpointMark.watermark;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public AmqpMessage getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public UnboundedAmqpSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      Read spec = source.spec;
+      messenger = Messenger.Factory.create();
+      messenger.start();
+      for (String address : spec.addresses()) {
+        messenger.subscribe(address);
+      }
+      checkpointMark.messenger = messenger;
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+      messenger.recv();
+      if (messenger.incoming() <= 0) {
+        current = null;
+        return false;
+      }
+      Message message = messenger.get();
+      Tracker tracker = messenger.incomingTracker();
+      checkpointMark.trackers.add(tracker);
+      currentTimestamp = new Instant(message.getCreationTime());
+      if (currentTimestamp.isBefore(checkpointMark.watermark)) {
+        checkpointMark.watermark = currentTimestamp;
+      }
+      current = new AmqpMessage(message);
+      return true;
+    }
+
+    @Override
+    public void close() {
+      if (messenger != null) {
+        messenger.stop();
+      }
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to send messages using AMQP 1.0 protocol.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<AmqpMessage>, PDone> {
+
+    @Nullable abstract List<String> addresses();
+    @Nullable abstract String subject();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setAddresses(List<String> addresses);
+      abstract Builder setSubject(String subject);
+      abstract Write build();
+    }
+
+    /**
+     * Define the AMQP addresses where to send messages.
+     */
+    public Write withAddresses(List<String> addresses) {
+      checkArgument(addresses != null, "AmqpIO.write().withAddresses(addresses) called with "
+          + "null addresses");
+      checkArgument(!addresses.isEmpty(), "AmpqIO.write().withAddresses(addresses) called with "
+          + "empty addresses list");
+      return builder().setAddresses(addresses).build();
+    }
+
+    /**
+     * Define the AMQP messages subject.
```

Comment:
Why do you need this? The user can already set the subject on the Message in the PCollection.